Amazon Bedrock - Knowledge Bases and Data Sources

ferruzzi opened the pull request:
Adds hooks, operators, sensors, triggers, and waiters to enable support for Amazon Bedrock Knowledge Bases.  This includes Bedrock Data Sources and Ingestion Jobs, and some light additions (a thin hook, and a waiter/sensor/trigger set) for Amazon OpenSearch Serverless which are needed to make the system test work end-to-end.

System test and unit tests are included, as well as docs.  Static checks, unit tests, and build-docs all pass locally.


<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 -->

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->



<!-- Please keep an empty line above the dashes. -->
---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

ferruzzi committed:
Amazon Bedrock Agent hook and unit tests

ferruzzi committed:
AOSS hook implemented

ferruzzi committed:
AOSS collection_available waiter/sensor/trigger with tests

ferruzzi committed:
create_data_source - operator and tests - takes under a second, no waiters

ferruzzi committed:
Update previous sensor unit tests to match new pattern

ferruzzi committed:
ingest_data - operator, trigger, sensor, and waiter (with unit tests)

ferruzzi committed:
Amazon Bedrock Knowledge Base system test

ferruzzi committed:
create_kb - operator, trigger, sensor, and waiter (with unit tests)

ferruzzi committed:
Doc pages - OpenSearch Serverless

ferruzzi committed:
Doc Page - Bedrock Agent

ferruzzi committed:
static check doc fixes

vincbeck commented on airflow/providers/amazon/aws/operators/bedrock.py:
```diff
@@ -351,3 +353,288 @@ def execute_complete(self, context: Context, event: dict[str, Any] | None = None
+        embedding_model_arn: str,
+        role_arn: str,
+        storage_config: dict[str, Any],
+        create_knowledge_base_kwargs: dict[str, Any] | None = None,
```
Missing from docstring

vincbeck commented on airflow/providers/amazon/aws/operators/bedrock.py:
```diff
@@ -351,3 +353,288 @@ def execute_complete(self, context: Context, event: dict[str, Any] | None = None
+        name: str,
+        knowledge_base_id: str,
+        bucket_name: str | None = None,
+        create_data_source_kwargs: dict[str, Any] | None = None,
```
Same

vincbeck commented on airflow/providers/amazon/aws/operators/bedrock.py:
```diff
@@ -351,3 +353,288 @@ def execute_complete(self, context: Context, event: dict[str, Any] | None = None
+        self,
+        knowledge_base_id: str,
+        data_source_id: str,
+        ingest_data_kwargs: dict[str, Any] | None = None,
```
Ok I stop :)

vincbeck commented on airflow/providers/amazon/aws/operators/bedrock.py:
```diff
@@ -351,3 +353,288 @@ def execute_complete(self, context: Context, event: dict[str, Any] | None = None
+            knowledgeBaseId=self.knowledge_base_id, dataSourceId=self.data_source_id
+        )["ingestionJob"]["ingestionJobId"]
+
+        if self.wait_for_completion:
```
No deferrable option for this one?

ferruzzi replied:
Good catch, I made the Trigger and forgot to implement it there.  Fix coming.

ferruzzi replied:
implemented

vincbeck commented on airflow/providers/amazon/aws/sensors/opensearch_serverless.py:
```diff
@@ -0,0 +1,110 @@
+    :param collection_id: A collection ID. You can't provide a name and an ID in the same request.
+    :param collection_name: A collection name. You can't provide a name and an ID in the same request.
+
+    :param deferrable: If True, the sensor will operate in deferrable more. This mode requires aiobotocore
+        module to be installed.
+        (default: False, but can be overridden in config file by setting default_deferrable to True)
```
I could not find this parameter in this class neither in its parent classes. And you dont seem to use it in the implementation. Left over I guess.

ferruzzi replied:
Ah, no, I made the trigger for that and forgot to implement it there.  Good catch, I'll add it now.

ferruzzi replied:
Implemented.

ferruzzi committed:
fix missing docstrings

ferruzzi committed:
implement ingestion job deferrable

ferruzzi committed:
add reference to opensearchserverless.rst

should fix "Validate provider.yaml files" failure

ferruzzi commented on airflow/providers/amazon/aws/sensors/opensearch_serverless.py:
```diff
@@ -0,0 +1,110 @@
+        collection_id: str | None = None,
+        collection_name: str | None = None,
+        poke_interval: int = 10,
+        max_retries: int = 120,
```
This default doesn't match the docstring; fix incoming.

ferruzzi committed:
Fix a new trigger bug and improve new trigger test coverage

ferruzzi committed:
more reasonable defaults for knowledge_base_active waiter

also fix

ferruzzi committed:
fix create knowledge base deferrable mode returning the wrong value

ferruzzi commented on airflow/providers/amazon/aws/sensors/bedrock.py:
```diff
@@ -68,7 +70,56 @@ def __init__(
+    def get_state(self) -> str:
+        """Implement in subclasses."""
+
+
```
@Taragolis  - These two base sensors are identical other than the aws_hook_class (BedrockHook vs BedrockAgentHook) and the hook name in the parent class.  I'm not really familiar with how you did the `AwsBaseSensor[BedrockAgentHook]` thing, is it possible to combine these into one BaseBedrockSensor?  What would that look like?

Taragolis replied:
If you would like to make generic from the generic you could do something like that

```python

_GenericBedrockHook = TypeVar("_GenericBedrockHook", BedrockAgentHook, BedrockHook)

class BedrockBaseSensor(AwsBaseSensor[_GenericBedrockHook]):
    """
    General sensor behavior for Amazon Bedrock.

    Subclasses must implement following methods:
        - ``get_state()``

    Subclasses must set the following fields:
        - ``INTERMEDIATE_STATES``
        - ``FAILURE_STATES``
        - ``SUCCESS_STATES``
        - ``FAILURE_MESSAGE``

    :param deferrable: If True, the sensor will operate in deferrable mode. This mode requires aiobotocore
        module to be installed.
        (default: False, but can be overridden in config file by setting default_deferrable to True)
    """

    INTERMEDIATE_STATES: tuple[str, ...]
    FAILURE_STATES: tuple[str, ...]
    SUCCESS_STATES: tuple[str, ...]
    FAILURE_MESSAGE: str

    ui_color = "#66c3ff"
    aws_hook_class: type[_GenericBedrockHook]

    ...

class BedrockCustomizeModelCompletedSensor(BedrockBaseSensor[BedrockHook]):
    aws_hook_class = BedrockHook
    ...

class BedrockKnowledgeBaseActiveSensor(BedrockBaseSensor[BedrockAgentHook]):
    aws_hook_class = BedrockAgentHook
    ...
```

ferruzzi replied:
AH!  I see, thanks.    I think I like the idea, but maybe I'll do that in a separate PR.  What do you think?  Does it feel worth doing?  Does it have to be now or can we let it stand how it is for now?

Taragolis replied:
I do not have personal preferences, so it could be done by the separate PR or even keep as it is.

ferruzzi committed:
Implement deferrable in OpenSearchServerlessCollectionActiveSensor

vincbeck reviewed (approved):
Beside the unit test failure you mentioned, LGTM!

ferruzzi committed:
account_id is no longer used, should have been cleaned up earlier

ferruzzi committed:
Move some hooks into tasks

ferruzzi committed:
fix opensearch unit test

This constant was removed because it was causing a circular import and I missed removing the references to it in the unit tests.

ferruzzi committed:
Add BedrockAgentBaseSensor to BASE_CLASSES

ferruzzi committed:
Standardize system test dag declaration

Taragolis commented on tests/system/providers/amazon/aws/example_bedrock_knowledge_base.py:
```diff
@@ -0,0 +1,519 @@
+    env_id = test_context["ENV_ID"]
+
+    aoss_client = OpenSearchServerlessHook().conn
+    bedrock_agent_client = BedrockAgentHook().conn
```
This two lines responses for calls to the DB. If you would like to avoid it you need to move it inside of the tasks, where it required.

Or if you intends to use only default `botocore` strategy you might explicitly specify it by provide `aws_conn_id=None`

ferruzzi replied:
They both get used a few times so I didn't want to recreate them multiple times.  Will the `aws_con_id=None` skip the db call and fix the static check?  For a system test, I think that's a fine solution.  I can try it out and see.  Thanks.

ferruzzi replied:
Looks like that did it.  Cheers.

Taragolis replied:
Yeah, `aws_con_id=None` skip get connection from the DB / env variable / secrets backends and only use configuration from hook

ferruzzi committed:
aws_conn_id=None

syedahsn commented on airflow/providers/amazon/aws/sensors/bedrock.py:
```diff
@@ -211,3 +262,149 @@ def execute(self, context: Context) -> Any:
+                    waiter_max_attempts=self.max_retries,
+                    aws_conn_id=self.aws_conn_id,
+                ),
+                method_name="poke",
```
Why is the callback being set to `poke`? Shouldn't it be set to some method that evaluates whatever the Trigger returns (e.g. the `event` object)?

ferruzzi replied:
I think this makes sense.   I could write another method that will check the event state, but poke is already doing that.  It results in one extra API call (the extra poke) but makes the code much nicer.   If you look at the execute_complete methods int he other sensors, all they do is `print state and if state != "success" then raise an exception`, that's just a copy/paste subset of the existing poke method.

ferruzzi replied:
if I modify poke from

```
  def poke(self, context: Context, **kwargs) -> bool:
        state = self.get_state()
        if state in self.FAILURE_STATES:
            # TODO: remove this if block when min_airflow_version is set to higher than 2.7.1
            if self.soft_fail:
                raise AirflowSkipException(self.FAILURE_MESSAGE)
            raise AirflowException(self.FAILURE_MESSAGE)

        return state not in self.INTERMEDIATE_STATES
 ```
 
 to (untested)
 
 
 ```
   def poke(self, context: Context, **kwargs) -> bool:
        state = kwargs.get("event").get("state") or self.get_state()
        if state in self.FAILURE_STATES or state == "failure":
            # TODO: remove this if block when min_airflow_version is set to higher than 2.7.1
            if self.soft_fail:
                raise AirflowSkipException(self.FAILURE_MESSAGE)
            raise AirflowException(self.FAILURE_MESSAGE)

        return state not in self.INTERMEDIATE_STATES
 ```       
 
 Then that would be the same thing without the extra get_state() call

syedahsn replied:
That makes sense to me. I think the extra call is worth the clarity. One thing I noticed is that `poke` doesn't accept `event` as a parameter. I'm not sure how important it is the the functionality (I'm sure you've tested it), but the docs seem to strongly recommend it:
>When your operator resumes, Airflow adds a context object and an event object to the kwargs passed to the method_name method. This event object contains the payload from the trigger event that resumed your operator. Depending on the trigger, this can be useful to your operator, like it’s a status code or URL to fetch results. Or, it might be unimportant information, like a datetime. Your method_name method, however, _must_ accept context and event as a keyword argument.

I don't feel too strongly about it either way. 
LGTM

ferruzzi replied:
Yeah, I initially added it as an explicit parameter but it raised issues with the static checks for not matching the signature of super().  Adding `**kwargs` passed the tests and `event` will just get swallowed up in there.  Since I'm not actually using it, I thought it was acceptable and in the end I decided it actually looked less confusing this way (as opposed to seeing 'event" there and wondering what you are supposed to give it if you aren't familiar with deferrables)


Are you alright with it how it is?  Maybe there's another way around that?  I can look if you care.  If you are alright with it how it is, can you resolve this convo?

syedahsn replied:
I'm good with it as is. Just wanted to make sure it was a conscious decision as opposed to an oversight.

syedahsn replied:
I don't see the option to resolve the conversation...

ferruzzi committed:
Fix templating

- The template fields are not substituted yet in the init so call_args was preserving and using the template string rather than the substituted value.
- Some of the names became too long once our env_id was appended, so I shortened them down.

ferruzzi commented on tests/system/providers/amazon/aws/example_bedrock_knowledge_base.py:
```diff
@@ -0,0 +1,519 @@
+# Licensed to the Apache Software Foundation (ASF) under one
```
Some of this needs adjusting, templated fields aren't working right with the way I wrote this.  Fix is coming,

ferruzzi replied:
done

ferruzzi committed:
more generic BedrockBaseSensor

ferruzzi committed:
remove now-non-existant base sensor
